source/particle/world.cc

```diff
@@ -531,31 +521,35 @@ namespace aspect
 
     template <int dim>
     void
-    World<dim>::sort_particles_in_subdomains_and_cells()
+    World<dim>::sort_particles_in_subdomains_and_cells(const std::vector<std::pair<types::LevelInd, Particle<dim> > > &particles_to_sort)
     {
       // TODO: The current algorithm only works for CFL numbers <= 1.0,
       // because it only knows the subdomain_id of ghost cells, but not
       // of artificial cells.
 
       // There are three reasons why a particle is not in its old cell:
-      // It moved to another cell, to another domain or it left the mesh.
-      // Sort the particles accordingly and deal with them
-      std::multimap<types::LevelInd, Particle<dim> >     moved_particles_cell;
-      std::multimap<types::subdomain_id, Particle<dim> > moved_particles_domain;
-      std::multimap<types::LevelInd, Particle<dim> >     lost_particles;
+      // It moved to another cell, to another subdomain or it left the mesh.
+      // Particles that moved to another cell are updated and stored inside the
+      // sorted_particles vector, particles that moved to another domain are
+      // collected in the moved_particles_domain multimap. Particles that left
+      // the mesh completely are collected in the lost_particles vector.
+      std::vector<std::pair<types::LevelInd, Particle<dim> > > sorted_particles;
+      std::multimap<types::subdomain_id, Particle<dim> >       moved_particles_domain;
+      std::vector<std::pair<types::LevelInd, Particle<dim> > > lost_particles;
+
+      sorted_particles.reserve(particles_to_sort.size()*2);
+      lost_particles.reserve(particles_to_sort.size());
```

Comment:
Sure I can do that. `lost_particles` really should only contain very few particles.